### ISSUE TITLE
几个疑惑

### ISSUE THREAD

**Issue by kscp123** [zceehua/Deep_Nature_Language_Processing#1]
多谢作者贡献tf版代码，在阅读时有几个地方不是很理解，望指点
这个函数里这几句不是很理解起到什么作用 
def gen_data(self,X):
        # bptt is seq len:70
        bptt = args.bptt if np.random.random() < 0.95 else args.bptt / 2.
        # Prevent excessively small or negative sequence lengths
        seq_len = max(5, int(np.random.normal(bptt, 5)))  # 每次抽取动态的seq len
        # There's a very small chance that it could select a very long sequence length resulting in OOM
        seq_len = min(seq_len, args.bptt + args.max_seq_len_delta)
还有embedding里的f1这步操作
 def f1():
            #drop word
            mask=tf.reshape(tf.multinomial(tf.log([[args.dropoute,1-args.dropoute]]),self.vocab_size),[self.vocab_size,1])
            mask=tf.cast(tf.tile(mask,[1,args.emsize]),tf.float32)#need to convert to float if we want consistent of data type
            self.embedding=self.embedding*mask
            return self.embedding
另外看到里面有一个py文件是rnn源码的修改，为什么要修改这个，改了哪里

不好意思一下提了这么多问题，本人菜鸟请见谅

**Comment by zceehua** [comment]
你好！
1.对于gen_data()函数，个人理解是原作者为了增加句子长短的多样性，使模型能够学习到在不同句长下的内容表达。主要是按照正太分布来取句子长短，bptt是我们自己预先设定好的平均句长。假如我们自定义的bptt很小，那么就规定最小句长为5（max），如果bptt很大，则规定最大句长为bptt+max_seq_len_delta（min）。其实也就是一个输入句长规则而已。
2.对于f1()函数，原作者代码中说**训练时**要对word取locked dropout， 也就是把单词按照一定百分比扔掉，所谓扔掉，其实就是把某些单词的embedding全部改成0.
3.个人比较懒直接把源码都粘贴过来了，其实只加了一行代码..，见1175行‘self._weights=tf.nn.dropout(self._weights,keep_prob=nnargs.wdrop)’，简单实现的drop connect链接。因为原论文中好像提到了LSTM之间要有drop connect。
大部分代码是按照源代码中改编而来，如果发现有错误，可以再联系。
谢谢:)

**Comment by kscp123** [comment]
多谢解答，这里有篇paper也说了这个问题：https://arxiv.org/pdf/1708.02182.pdf  4.1节，基本就是你说的那样。另外args里面那个drop的prob值是不是都应该用1减去啊，tf里的是keep也就是保留多少。如果是keep 0应该是drop 100%吧

**Comment by kscp123** [comment]
@zceehua 请问作者跑这个用的什么配置？我这11g的K80还是报ResourceExhaustedError。。

**Comment by zceehua** [comment]
我之前在8G的1080上也有过这个问题，这个模型原来默认参数设定都很大，后来在24G的1080上跑的并且稍微调小了些参数。把batch size或者其他参数弄小一点试试看。目前除了用更多GPU外我也不太清楚如何解决。
